### PR TITLE
Remove locking on shared blob cache service for hot pages

### PR DESCRIPTION
**Pull request by original-brownbear**
Locking on the cache service is unnecessary, the access time can be safely checked through a volatile read since it's monotonically increasing. We will never not pass the check outside the lock here and then pass it inside the lock with the fixed `now`.
If another thread concurrently promotes the entry, then the time check under the lock will practically always fail and without a concurrent update to the entry it doesn't matter in the first place.
-> This removes the non-scalable (in number of threads) global lock for pages that are hot most of the time, which should do away with some tail latency that we see in benchmarks.

### CODE REVIEW

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java [hunk 1]
```diff
@@ -602,12 +608,10 @@ private void computeDecay() {
             long now = threadPool.relativeTimeInMillis();
             for (int i = 0; i < maxFreq; i++) {
                 for (Entry<CacheFileRegion> entry = freqs[i]; entry != null; entry = entry.next) {
-                    if (now - entry.lastAccessed >= 2 * minTimeDelta) {
-                        if (entry.freq > 0) {
-                            unlink(entry);
-                            entry.freq--;
-                            pushEntryToBack(entry);
-                        }
+                    if (entry.freq > 0 && now - entry.lastAccessed >= 2 * minTimeDelta) {
```
Comment (original-brownbear) [inline x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java:611]: Bit of vodoo, the `freq` is not volatile, so lets check it first to save some cycles here and there, also the needlessly nested `if` was annoying :)

**Review by romseygeek — APPROVED** [review]
Nice!

**Comment by original-brownbear** [comment]
Thanks Alan!

**Review by tlrx — COMMENTED** [review]
LGTM

**Pull request merged**